### PR TITLE
Fix #4914 patternize filter overrides

### LINKED ISSUE #4914: Consider patternizing concern overrides for filters
Code sample from https://github.com/aspnet/Mvc/issues/4506#issuecomment-229093334

```
public interface IFooConcernFilter
{
    // Just a marker
}

public class MyGlobalFooConcernFilter : IFooConcernFilter, IActionFilter
{
    public void OnActionExecuting(...)
    {
        if (!IsClosestFooConcernFilter(context)
        {
            return;
        }
    }

    private bool IsClosestFooConcernFilter(FilterContext context)
    {
            // Determine if this instance is the 'effective' foo concern policy.
            for (var i = filters.Count - 1; i >= 0; i--)
            {
                var filter = filters[i];
                if (filter is IFooConcernFilter)
                {
                    return object.ReferenceEquals(this, filter);
                }
            }

            Debug.Fail("The current instance should be in the list of filters.");
            return false;
    }
}

public class IgnoreFooConcernAttribute : Attribute, IFooConcernFilter
{
    // Also just a marker
}
```

This gives `[IgnoreFooConcern]` overriding behavior. If you slap that on a controller or action, then the global filter knows to shut itself off.

We do this kind of thing frequently in the framework, and it's a non-obvious pattern. Maybe we should add some kind of infrastructure/extensions to make this easier for users to grok.

### PR DESCRIPTION
**Pull request by rynowak**
Adding some convenience methods to the context class and updating our
code to use them.

### CODE REVIEW

**Comment by rynowak** [comment]
Turns out we had 5 implementations of this pattern

File: src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/FilterContext.cs [hunk 1]
```diff
@@ -33,5 +33,67 @@ public abstract class FilterContext : ActionContext
...
+        /// The policy applied directly to the action method could be considered more specific.
+        /// </para>
+        /// <para>
+        /// This mechanism for overriding relies and the rules of order and scope that the filter system
+        /// provides to control ordering of filters. It is up to the implementor of filters to implement this 
+        /// protocol cooperatively. The filter system has no innate notion of overrides, this is a recommended
+        /// convention.
+        /// </para>
+        /// </remarks>
+        public bool IsEffectivePolicy<TMetadata>(TMetadata policy) where TMetadata : IFilterMetadata
```
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/FilterContext.cs]: A user would still need to check for this explicitly though. Could we introduce a new marker interface for this scenario (something like IClosestFilterWins or something)? So a user writing a filter can just implement this marker interface and we take care of the logic internally and pass around the 'final' list of filters through a new property like 'EffectiveFilters' or something on the FilterContext?

Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/FilterContext.cs]: No, I want to keep this as a convention. We decided early on not to build filter overrides into the system

Comment (kichalla) [reply src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/FilterContext.cs]: Sure 👍 . Just wondering if finding the effective filter should be driven by 'policy' alone. For example, could we provide an additional overload which checks the current type of the filter alone. This way a user doesn't need to define a new type for 'policy'.

File: src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/FilterContext.cs [hunk 2]
```diff
@@ -33,5 +33,67 @@ public abstract class FilterContext : ActionContext
...
+        /// cross-cutting concern, define a common interface for the filters (<typeparamref name="TMetadata"/>) and 
+        /// implement the filters such that all of the implementations call this method to determine if they should
+        /// take action.
+        /// </para>
+        /// <para>
+        /// For instance, a global filter might be overridden by placing a filter attribute on an action method.
+        /// The policy applied directly to the action method could be considered more specific.
+        /// </para>
+        /// <para>
+        /// This mechanism for overriding relies and the rules of order and scope that the filter system
```
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/FilterContext.cs]: `relies on`

File: src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/FilterContext.cs [hunk 3]
```diff
@@ -33,5 +33,67 @@ public abstract class FilterContext : ActionContext
...
+        /// </summary>
+        /// <typeparam name="TMetadata">The type of the filter policy.</typeparam>
+        /// <param name="policy">The filter policy instance.</param>
+        /// <returns>
+        /// <c>true</c> if the provided <see cref="IFilterMetadata"/> is the most effective policy, otherwise <c>false</c>.
+        /// </returns>
+        /// <remarks>
+        /// <para>
+        /// The <see cref="IsEffectivePolicy{TMetadata}(TMetadata)"/> method is used to implement a common convention
+        /// for filters where when that define an overriding behavior. When multiple filters may apply to the same
```
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/FilterContext.cs]: remove `where when`?

**Pull request merged**